File: 2024/day03/zig/solution.zig
const std = @import("std");

const Event = struct {
    pos: usize,
    kind: enum { mul, do_enable, dont_disable },
    x: i32,
    y: i32,
};

fn lessThan(_: void, a: Event, b: Event) bool {
    return a.pos < b.pos;
}

fn part1(data: []const u8) !i32 {
    var total: i32 = 0;
    var i: usize = 0;

    while (i < data.len) {
        // Look for "mul("
        if (i + 4 < data.len and std.mem.eql(u8, data[i..i+4], "mul(")) {
            var j = i + 4;

            // Parse first number (1-3 digits)
            var x: i32 = 0;
            var digit_count: usize = 0;
            while (j < data.len and digit_count < 3 and std.ascii.isDigit(data[j])) {
                x = x * 10 + @as(i32, data[j] - '0');
                j += 1;
                digit_count += 1;
            }

            // Must have at least 1 digit
            if (digit_count == 0) {
                i += 1;
                continue;
            }

            // Must have comma
            if (j >= data.len or data[j] != ',') {
                i += 1;
                continue;
            }
            j += 1;

            // Parse second number (1-3 digits)
            var y: i32 = 0;
            digit_count = 0;
            while (j < data.len and digit_count < 3 and std.ascii.isDigit(data[j])) {
                y = y * 10 + @as(i32, data[j] - '0');
                j += 1;
                digit_count += 1;
            }

            // Must have at least 1 digit
            if (digit_count == 0) {
                i += 1;
                continue;
            }

            // Must have closing paren
            if (j >= data.len or data[j] != ')') {
                i += 1;
                continue;
            }

            // Valid mul instruction!
            total += x * y;
            i = j + 1;
        } else {
            i += 1;
        }
    }

    return total;
}

fn part2(allocator: std.mem.Allocator, data: []const u8) !i32 {
    var events = try std.ArrayList(Event).initCapacity(allocator, 100);
    defer events.deinit(allocator);

    var i: usize = 0;

    // Find all mul instructions
    while (i < data.len) {
        if (i + 4 < data.len and std.mem.eql(u8, data[i..i+4], "mul(")) {
            var j = i + 4;

            // Parse first number (1-3 digits)
            var x: i32 = 0;
            var digit_count: usize = 0;
            while (j < data.len and digit_count < 3 and std.ascii.isDigit(data[j])) {
                x = x * 10 + @as(i32, data[j] - '0');
                j += 1;
                digit_count += 1;
            }

            if (digit_count == 0 or j >= data.len or data[j] != ',') {
                i += 1;
                continue;
            }
            j += 1;

            // Parse second number (1-3 digits)
            var y: i32 = 0;
            digit_count = 0;
            while (j < data.len and digit_count < 3 and std.ascii.isDigit(data[j])) {
                y = y * 10 + @as(i32, data[j] - '0');
                j += 1;
                digit_count += 1;
            }

            if (digit_count == 0 or j >= data.len or data[j] != ')') {
                i += 1;
                continue;
            }

            // Valid mul instruction!
            try events.append(allocator, Event{
                .pos = i,
                .kind = .mul,
                .x = x,
                .y = y,
            });
            i = j + 1;
        } else {
            i += 1;
        }
    }

    // Find all do() instructions
    i = 0;
    while (i < data.len) {
        if (i + 4 <= data.len and std.mem.eql(u8, data[i..i+4], "do()")) {
            try events.append(allocator, Event{
                .pos = i,
                .kind = .do_enable,
                .x = 0,
                .y = 0,
            });
            i += 4;
        } else {
            i += 1;
        }
    }

    // Find all don't() instructions
    i = 0;
    while (i < data.len) {
        if (i + 7 <= data.len and std.mem.eql(u8, data[i..i+7], "don't()")) {
            try events.append(allocator, Event{
                .pos = i,
                .kind = .dont_disable,
                .x = 0,
                .y = 0,
            });
            i += 7;
        } else {
            i += 1;
        }
    }

    // Sort events by position
    std.mem.sort(Event, events.items, {}, lessThan);

    // Process events in order
    var total: i32 = 0;
    var enabled = true;

    for (events.items) |event| {
        switch (event.kind) {
            .do_enable => enabled = true,
            .dont_disable => enabled = false,
            .mul => {
                if (enabled) {
                    total += event.x * event.y;
                }
            },
        }
    }

    return total;
}

pub fn main() !void {
    var gpa = std.heap.GeneralPurposeAllocator(.{}){};
    defer _ = gpa.deinit();
    const allocator = gpa.allocator();

    // Read input file
    const file = try std.fs.cwd().openFile("../input.txt", .{});
    defer file.close();

    const data = try file.readToEndAlloc(allocator, 1024 * 1024);
    defer allocator.free(data);

    const p1 = try part1(data);
    const p2 = try part2(allocator, data);

    std.debug.print("Part 1: {d}\n", .{p1});
    std.debug.print("Part 2: {d}\n", .{p2});
}

File: 2024/day03/rust/solution.rs
use std::fs;

fn part1(data: &str) -> i32 {
    let mut total = 0;
    let bytes = data.as_bytes();
    let mut i = 0;

    while i < bytes.len() {
        // Look for "mul("
        if i + 4 < bytes.len()
            && bytes[i] == b'm'
            && bytes[i+1] == b'u'
            && bytes[i+2] == b'l'
            && bytes[i+3] == b'(' {

            i += 4; // Move past "mul("

            // Parse first number (1-3 digits)
            let mut x = 0;
            let mut digit_count = 0;
            while i < bytes.len() && bytes[i].is_ascii_digit() && digit_count < 3 {
                x = x * 10 + (bytes[i] - b'0') as i32;
                i += 1;
                digit_count += 1;
            }

            // Must have at least 1 digit
            if digit_count == 0 {
                continue;
            }

            // Must have comma
            if i >= bytes.len() || bytes[i] != b',' {
                continue;
            }
            i += 1; // Move past comma

            // Parse second number (1-3 digits)
            let mut y = 0;
            digit_count = 0;
            while i < bytes.len() && bytes[i].is_ascii_digit() && digit_count < 3 {
                y = y * 10 + (bytes[i] - b'0') as i32;
                i += 1;
                digit_count += 1;
            }

            // Must have at least 1 digit
            if digit_count == 0 {
                continue;
            }

            // Must have closing paren
            if i < bytes.len() && bytes[i] == b')' {
                total += x * y;
                i += 1; // Move past closing paren
            }
        } else {
            i += 1;
        }
    }

    total
}

fn part2(data: &str) -> i32 {
    let bytes = data.as_bytes();
    let mut i = 0;
    let mut total = 0;
    let mut enabled = true;

    while i < bytes.len() {
        // Check for "do()"
        if i + 4 <= bytes.len()
            && bytes[i] == b'd'
            && bytes[i+1] == b'o'
            && bytes[i+2] == b'('
            && bytes[i+3] == b')' {
            enabled = true;
            i += 4;
            continue;
        }

        // Check for "don't()"
        if i + 7 <= bytes.len()
            && bytes[i] == b'd'
            && bytes[i+1] == b'o'
            && bytes[i+2] == b'n'
            && bytes[i+3] == b'\''
            && bytes[i+4] == b't'
            && bytes[i+5] == b'('
            && bytes[i+6] == b')' {
            enabled = false;
            i += 7;
            continue;
        }

        // Check for "mul("
        if i + 4 < bytes.len()
            && bytes[i] == b'm'
            && bytes[i+1] == b'u'
            && bytes[i+2] == b'l'
            && bytes[i+3] == b'(' {

            let start = i;
            i += 4; // Move past "mul("

            // Parse first number (1-3 digits)
            let mut x = 0;
            let mut digit_count = 0;
            while i < bytes.len() && bytes[i].is_ascii_digit() && digit_count < 3 {
                x = x * 10 + (bytes[i] - b'0') as i32;
                i += 1;
                digit_count += 1;
            }

            // Must have at least 1 digit
            if digit_count == 0 {
                i = start + 1;
                continue;
            }

            // Must have comma
            if i >= bytes.len() || bytes[i] != b',' {
                i = start + 1;
                continue;
            }
            i += 1; // Move past comma

            // Parse second number (1-3 digits)
            let mut y = 0;
            digit_count = 0;
            while i < bytes.len() && bytes[i].is_ascii_digit() && digit_count < 3 {
                y = y * 10 + (bytes[i] - b'0') as i32;
                i += 1;
                digit_count += 1;
            }

            // Must have at least 1 digit
            if digit_count == 0 {
                i = start + 1;
                continue;
            }

            // Must have closing paren
            if i < bytes.len() && bytes[i] == b')' {
                if enabled {
                    total += x * y;
                }
                i += 1; // Move past closing paren
            } else {
                i = start + 1;
            }
        } else {
            i += 1;
        }
    }

    total
}

fn main() {
    let data = fs::read_to_string("../input.txt")
        .expect("Failed to read input file");

    println!("Part 1: {}", part1(&data));
    println!("Part 2: {}", part2(&data));
}
